arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Bracket.kt

```diff
@@ -5,21 +5,26 @@ import kotlinx.coroutines.CancellationException
 import kotlinx.coroutines.NonCancellable
 import kotlinx.coroutines.withContext
 
-public sealed class ExitCase {
-  public object Completed : ExitCase() {
-    override fun toString(): String =
-      "ExitCase.Completed"
-  }
-
-  public data class Cancelled(val exception: CancellationException) : ExitCase()
-  public data class Failure(val failure: Throwable) : ExitCase()
+public sealed interface ExitCase {
+  public data object Completed : ExitCase
+  public data class Cancelled(val exception: CancellationException) : ExitCase
+  public data class Failure(val failure: Throwable) : ExitCase
 
   public companion object {
-    public fun ExitCase(error: Throwable): ExitCase =
-      if (error is CancellationException) Cancelled(error) else Failure(error)
+    public fun ExitCase(error: Throwable?): ExitCase = when (error) {
+      null -> Completed
+      is CancellationException -> Cancelled(error)
+      else -> Failure(error)
+    }
   }
 }
 
+public val ExitCase.throwableOrNull: Throwable? get() = when (this) {
+  ExitCase.Completed -> null
+  is ExitCase.Cancelled -> exception
+  is ExitCase.Failure -> failure
+}
+
```

Comment:
Can we define this inside `ExitCase` so that we don't need an import for it? Preferably as a function I'd say, that'd be more in line with the APIs we now from `Result`.

I also think `exceptionOrNull` is a more common name, than `throwableOrNull`.